Prepare GraalSOM for partial evaluation

sophie-kaleba opened the pull request:
It mainly makes the language more visible to the framework, notably by relying on `CallTarget` and the `execute` method.
The frame handling responsibility has been transferred out of the `Interpreter` to likely help for a future switch to Truffle `Frame`.

@smarr - I have added a test case about binary operations (where each operand is a message send, the test looks roughly like this: `^ (self foo: 1) + (self foo2: 2)`). Is it OK to push it to the SOM repository? Also, could you have a look at this? Thanks!

smarr commented:
@sophie-kaleba sorry, the workflow with the submodules is a bit annoying.
Could you open an PR on https://github.com/SOM-st/SOM with the test?

smarr commented on src/som/interpreter/Frame.java:
```diff
@@ -227,6 +227,19 @@ public void printStackTrace(final SAbstractObject nilObject) {
   }
 
+  public void popArgumentsAndPushResult(final SAbstractObject result, SMethod method) {
+    // compute the number of arguments
```
These type of comments seem to be redundant with the code and don't give a rational.
So, whenever I edit these types of methods, I tent to remove them.

Though, you are having a fresh look at the code, do you find them useful?

sophie-kaleba replied:
No, indeed. At least not here. I've removed them

smarr commented on src/som/interpreter/Interpreter.java:
```diff
@@ -46,185 +66,171 @@ public Interpreter(final Universe universe) {
-      int numArgs = method.getNumberOfArguments();
-      for (int i = 0; i < numArgs; i += 1) {
-        getFrame().pop();
-      }
```
I am probably not yet fully awake, but do you still need to pop these arguments from the frame?

sophie-kaleba replied:
Hmm, I still need to have a look. I think it kind of works right now because the following `sendEscapedBlock` call will only fetch the arguments from the given method and will ignore the ones that have not been popped

smarr replied:
There might just not be a test that runs out of stack.
I'd be careful with this.

smarr commented on src/som/interpreter/Interpreter.java:
```diff
@@ -325,39 +329,22 @@ public SAbstractObject start() throws ProgramDefinitionError {
 
   private void send(final SSymbol selector, final SClass receiverClass,
-      final int bytecodeIndex) {
+      final int bytecodeIndex, Frame frame) throws ReturnException {
```
Hm, you don't throw a `ReturnException` here, do you?

sophie-kaleba replied:
True, removed.

smarr commented on src/som/interpreter/Interpreter.java:
```diff
@@ -383,44 +370,24 @@ private void send(final SSymbol selector, final SClass receiverClass,
-    // Pop the top frame from the interpreter frame stack and compute the
-    // number of arguments
-    int numberOfArguments = popFrame().getMethod().getNumberOfArguments();
+  private final IndirectCallNode indirectCallNode = Truffle.getRuntime().createIndirectCallNode();
```
Please order all fields at the top of the class (as per standard Java practice), I know, SOM does it at the bottom of the class, but I'd like to get away from that as well.

smarr commented on src/som/interpreter/Method.java:
```diff
@@ -0,0 +1,67 @@
+                programDefinitionError.printStackTrace();
+            }
+            return null;
+        }
```
This method won't pass CI and the formatting requirements.
Perhaps you can somehow setup InteliJ to do a format on save automatically, using the Eclipse formatting settings.
I am pretty sure there's a way, since the Graal/Truffle team seems to use it.

smarr commented on src/som/interpreter/ReturnException.java:
```diff
@@ -0,0 +1,26 @@
+
+import som.vmobjects.SAbstractObject;
+
+public class ReturnException extends RuntimeException {
```
You may already want to use Truffle's `ControlFlowException` class as superclass here.

smarr commented on src/som/vm/Shell.java:
```diff
@@ -69,8 +71,8 @@ public SAbstractObject start() {
 
-    // Create a fake bootstrap frame
-    currentFrame = interpreter.pushNewFrame(bootstrapMethod);
+    // fetch the bootstrap frame
```
Is this comment helpful? Perhaps rename `frame` to `bootstrapFrame` so it's explicit in the code?

smarr commented on src/som/vm/Shell.java:
```diff
@@ -94,7 +96,7 @@ public SAbstractObject start() {
         // If success
         if (myClass != null) {
-          currentFrame = interpreter.getFrame();
+          currentFrame = frame;
```
Hm, I am not entirely sure whether these two are equivalent.
The frame may have changed in the interpreter.

The question is whether we want/need to support this.

Is there a use case where we want the loading of a class to be able to change the frame?

We are also in the loop, so, on the next iteration, the code may have returned from another frame? Is that possible? I am not sure. The code is not explicit about why that's needed, is it?

sophie-kaleba replied:
I hadn't really paid attention to the shell mode to be honest. Now that I did, I tried to delete the silly hot fixes I added to make it compile. It should behave now the same way as it was behaving in som-java / TruffleSOM. This could be improved but I am not sure I want to spend time on this for now.

sophie-kaleba commented:
The PR for the new test case in SOM is there: https://github.com/SOM-st/SOM/pull/33

sophie-kaleba commented:
I have pushed changes that have been made during the last couple of weeks.
Basically:
- The interpreter loop now gets properly inlined: compilation constants in the loop (bytecode index, bytecode length, frame and method) have been annotated accordingly. 
- The `Frame` handling has been removed from within the interpreter
- Added a configuration file for benchmarks (still needs refining)

I also fixed the formatting, so sorry about the messy diffs...

smarr commented on src/som/interpreter/Bytecodes.java:
```diff
@@ -78,29 +80,35 @@ public static String getPaddedBytecodeName(byte bytecode) {
   }
 
-  public static int getBytecodeLength(byte bytecode) {
+  public static final int getBytecodeLength(final byte bytecode) {
```
Think `final` might not be necessary/redundant. And checkstyle would complain about it.

sophie-kaleba replied:
I explicitly "checkstyled" the whole project before committing and did not get any complaints. I'll check whether I am using the right files

sophie-kaleba replied:
actually, why should it complain? I checked my configuration and it seems similar to SOMns except the modified bits about tests/copyrights I had to remove to make intellij-checkstyle happy

smarr replied:
My assumption is that static methods cannot be overridden, which makes the final modifier redundant, I think.

sophie-kaleba replied:
Ha yes indeed

smarr commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,36 @@
+before_script:
+  - git submodule update --init
+
+# I'll disable it just for now, I haven't fixed checkstyle in my IDE yet
```
Did you try: https://plugins.jetbrains.com/plugin/1065-checkstyle-idea

If it crashes, you may need to update the checkstyle config. In SOMns, it should already be updated: https://github.com/smarr/SOMns/commit/684aa34b6cc397449b7b0e8fc1119914908fef9c

sophie-kaleba replied:
I should re-enable it actually, I installed this plugin a while back and it works. Some bits of the checkstyle configuration files from SOMns do not work with it (but on not-so-relevant bits such as formatting of tests/copyright headers...) but overall it does the job

smarr commented on src/som/interpreter/Frame.java:
```diff
@@ -217,25 +217,13 @@ public void copyArgumentsFrom(final Frame frame) {
     }
   }
 
-  public void printStackTrace(final SAbstractObject nilObject) {
```
Hm, you're already relying on an implicit stack? might still be useful for debugging.

smarr commented:
You want me to merge this now?

sophie-kaleba commented:
Let's postpone the merge until tomorrow to see if I can make the inlining of direct calls happen?

sophie-kaleba commented:
@smarr - it should be good for merging now. 
I have added the last changes for inline caches (refactoring of the code to give more hints for PE) and the use of ValueProfile to get more insight on the receiver class in sends. The inlining of direct calls does not work yet unfortunately.

smarr commented on src/som/vmobjects/SAbstractObject.java:
```diff
@@ -24,51 +24,45 @@
-
   public abstract SClass getSOMClass(Universe universe);
 
+  public abstract ValueProfile getValueProfile();
```
This seems very costly.
I would assume there are many many more run-time values than lexical locations. At least for peak-performance purposes. So, putting the value here seems like it would cause a very high overhead.

smarr commented on src/som/vmobjects/SDouble.java:
```diff
@@ -24,16 +24,21 @@
 public final class SDouble extends SNumber {
 
+  @CompilerDirectives.CompilationFinal private final ValueProfile valueProfile =
+      ValueProfile.createClassProfile();
```
Would drop this. This means that every single double number gets a separate profile. This seems very costly.

smarr commented on src/som/vmobjects/SInteger.java:
```diff
@@ -29,12 +29,19 @@
+
 public final class SInteger extends SNumber {
 
+  @CompilationFinal private final ValueProfile valueProfile =
```
Would drop this.

smarr commented on src/som/vmobjects/SMethod.java:
```diff
@@ -26,21 +26,48 @@
+  private final @CompilationFinal(dimensions = 1) SClass[]          inlineCacheClass;
+  private final @CompilationFinal(dimensions = 1) SInvokable[]      inlineCacheInvokable;
+  final @CompilationFinal(dimensions = 1) DirectCallNode[]          inlineCacheDirectCallNodes;
+  final @CompilationFinal(dimensions = 1) ValueProfile[]            receiverClassProfiles;
```
Yes! This is what we need. Exactly this!

smarr commented on src/som/vmobjects/SMethod.java:
```diff
@@ -26,21 +26,48 @@
+  final @CompilationFinal(dimensions = 1) DirectCallNode[]          inlineCacheDirectCallNodes;
+  final @CompilationFinal(dimensions = 1) ValueProfile[]            receiverClassProfiles;
+  private final @CompilationFinal(dimensions = 1) SAbstractObject[] literals;
+  private final @CompilationFinal ValueProfile                      valueProfile =
```
I'd just drop this.

smarr commented on src/som/interpreter/Interpreter.java:
```diff
@@ -25,297 +25,291 @@
+    ValueProfile receiverClassValueProfile = method.getValueProfile(bytecodeIndex);
+    if (receiverClassValueProfile == null) {
+      CompilerDirectives.transferToInterpreterAndInvalidate();
+      receiverClassValueProfile = receiver.getValueProfile();
```
I am not convinced this is a good idea, but now I see what you are doing.
I would recommend against this, and simply construct a profile here at this point.

smarr commented on src/som/vmobjects/SPrimitive.java:
```diff
@@ -61,11 +72,25 @@ public boolean isEmpty() {
     return universe.primitiveClass;
   }
 
+  public final ValueProfile getValueProfile() {
```
Would drop all these.

smarr commented on src/som/vmobjects/SSymbol.java:
```diff
@@ -24,17 +24,24 @@
 
+  private final int numberOfSignatureArguments;
+
+  @CompilerDirectives.CompilationFinal private final ValueProfile valueProfile =
```
Would drop.

smarr commented on src/som/interpreter/Interpreter.java:
```diff
@@ -25,297 +25,291 @@
+    if (receiverClassValueProfile == null) {
+      CompilerDirectives.transferToInterpreterAndInvalidate();
+      receiverClassValueProfile = ValueProfile.createClassProfile();
+      method.setValueProfile(bytecodeIndex, receiverClassValueProfile);
```
Would be nice to call this `setReceiverProfile`

smarr commented on src/som/vmobjects/SMethod.java:
```diff
@@ -26,21 +26,45 @@
-      int numberOfLiterals, List<SAbstractObject> literals) {
+  private final @CompilationFinal(dimensions = 1) SClass[]          inlineCacheClass;
+  private final @CompilationFinal(dimensions = 1) SInvokable[]      inlineCacheInvokable;
+  final @CompilationFinal(dimensions = 1) DirectCallNode[]          inlineCacheDirectCallNodes;
+  final @CompilationFinal(dimensions = 1) ValueProfile[]            receiverClassProfiles;
```
Should these be private?

smarr commented on src/som/vmobjects/SMethod.java:
```diff
@@ -130,35 +173,49 @@ public String toString() {
+    return inlineCacheDirectCallNodes[bytecodeIndex];
+  }
+
+  public void setValueProfile(final int bytecodeIndex,
```
Could we rename this to `setReceiverProfile`?
